### PR TITLE
docs: the blockchain, state tree, and snapshots

### PR DESCRIPTION
**Pull request by aatifsyed**
[rendered](https://github.com/ChainSafe/forest/files/12927509/rendered.pdf)

A ground-up education on:
- Actors
- The Filecoin blockchain
- The Filecoin state tree
- Snapshots

I didn't really understand how these all fit together until the past couple weeks.

# Reviewers
- Check for correctness please - there's a to-do in there for you also
- I thought account balances etc would be a more motivated example of state,
  but I can't find where those are in the graph.
  The [`fil_actor_account_state`](https://docs.rs/fil_actor_account_state/6.1.0/fil_actor_account_state/v11/struct.State.html)
  doesn't look how I expect at all.

# Location
I strongly believe (this) documentation should live in-tree, in `/src`, not `/documentation`.
It will:
- be able to link to the code and ecosystem.
- be able to link to other in-tree documentation.
- be more seen.

### CODE REVIEW

**Comment by aatifsyed** [comment]
[rendered](https://github.com/ChainSafe/forest/files/12927509/rendered.pdf)

File: src/lib.rs [hunk 1]
```diff
@@ -1,8 +1,230 @@
...
+//!
+//! An example of an actor is the [`cron`](fil_actors_shared::v11::runtime::builtins::Type::Cron)
+//! actor.
+//! Its [internal state](fil_actor_cron_state::v11::State) is a to-do list of
+//! other actors to invoke every epoch.
+//!
+//! See [the Filecoin docs](https://docs.filecoin.io/basics/the-blockchain/actors)
+//! for more information about actors.
+//!
+//! [^1]: The number of [built-in actors](https://docs.filecoin.io/basics/the-blockchain/actors#built-in-actors)
```
Comment (jdjaustin) [inline src/lib.rs:18]:
```suggestion
//! [^1]: The number of [built-in actors](https://github.com/filecoin-project/builtin-actors)
```
As you already linked to the Filecoin docs in the lines directly above, I think the link to the repo makes more sense here and (as an added bonus) is less likely to suffer from link rot.

File: src/lib.rs [hunk 2]
```diff
@@ -1,8 +1,230 @@
...
+//! TODO(aatifsyed): where is my balance?
+//!
+//! # The Filecoin blockchain
+//!
+//! Filecoin consists of a blockchain of messages.
+//! Listed below are the core objects for the blockchain.
+//! Each one can be addressed by a [`Cid`](cid::Cid).
+//!
+//! - [`Message`](shim::message::Message)s are statements of messages between
+//!   a the actors.
```
Comment (jdjaustin) [inline src/lib.rs:30]:
```suggestion
//!   the actors.
```

File: src/lib.rs [hunk 3]
```diff
@@ -1,8 +1,230 @@
...
+//!   They describe and (equivalently) represent a change in _the blockchain state_.
+//!   See [`apply_block_messages`](state_manager::apply_block_messages) to learn
+//!   more.
+//!   Messages may be [signed](message::SignedMessage).
+//!   TODO(aatifsyed): by whom? what does that mean?
```
Comment (jdjaustin) [inline src/lib.rs:35]: From https://spec.filecoin.io/systems/filecoin_vm/message/ :

"A semantically valid `SignedMessage` must carry a signature that verifies the payload as having been signed with the public key of the account actor identified by the `From` address. Note that when the `From` address is an ID-address, the public key must be looked up in the state of the sending account actor in the parent state identified by the block."

File: src/lib.rs (hunk 1 above)
Comment (aatifsyed) [reply src/lib.rs:18]: You're right - the redundant link isn't worth it, I'll just remove the link altogether though

**Review by lemmih — APPROVED** [review]

**Review by hanabi1224 — APPROVED** [review]

**Pull request merged**